crates/transpiler/src/passes/alap_schedule_analysis.rs

```diff
@@ -0,0 +1,177 @@
+            .cargs_interner()
+            .get(op.clbits)
+            .iter()
+            .map(|&c| Wire::Clbit(c))
+            .collect();
+
+        /*
+         compute t0, t1: instruction interval, note that
+         t0: start time of instruction
+         t1: end time of instruction
+
+         since this is alap scheduling, node is scheduled in reversed topological ordering
+         and nodes are packed from the very end of the circuit.
+         the physical meaning of t0 and t1 is flipped here.
+        */
+
+        // Get operation type
+        let op_name = op.op.name();
+        let op_view = op.op.view();
+        let (is_gate, is_delay) = match op_view {
+            OperationRef::Gate(_) => (true, false),
+            OperationRef::StandardGate(_) => (true, false),
+            _ => (false, op_name == "delay"),
+        };
+
+        let t1 = if is_gate || is_delay {
+            // Gate or Delay operation
+            let t0 = qargs
+                .iter()
+                .map(|q| *idle_before.get(q).unwrap_or(&0.0))
+                .max_by(|a, b| a.partial_cmp(b).unwrap())
+                .unwrap_or(0.0);
+            t0 + op_duration
+        } else if op_name == "measure" {
+            // Measure operation
+            // clbit time is always right (alap) justified
+            let t0 = qargs
+                .iter()
+                .chain(cargs.iter())
+                .map(|bit| *idle_before.get(bit).unwrap_or(&0.0))
+                .max_by(|a, b| a.partial_cmp(b).unwrap())
+                .unwrap_or(0.0);
+            /*
+                       |t1 = t0 + duration
+                Q ░░░░░▒▒▒▒▒▒▒▒▒▒▒
+                C ░░░░░░░░░▒▒▒▒▒▒▒
+                           |t0 + (duration - clbit_write_latency)
+            */
+            let t1 = t0 + op_duration;
+            for clbit in cargs.iter() {
+                idle_before.insert(*clbit, t0 + (op_duration - clbit_write_latency as f64));
+            }
+            t1
+        } else {
+            // Directive (like barrier)
+            let t0 = qargs
+                .iter()
+                .chain(cargs.iter())
+                .map(|bit| *idle_before.get(bit).unwrap_or(&0.0))
+                .max_by(|a, b| a.partial_cmp(b).unwrap())
+                .unwrap_or(0.0);
+            t0 + op_duration
+        };
+
+        for qubit in qargs {
+            idle_before.insert(qubit, t1);
+        }
+
+        node_start_time.insert(node_index, t1);
+    }
+
+    // Compute maximum instruction available time
+    let circuit_duration = *idle_before
+        .values()
+        .max_by(|a, b| a.partial_cmp(b).unwrap())
+        .unwrap_or(&0.0);
+    // Note that ALAP pass is inversely schedule, thus
+    // t0 is computed by subtracting entire circuit duration from t1.
+    let node_start_time_final: HashMap<NodeIndex, f64> = node_start_time
+        .into_iter()
+        .map(|(n, t1)| (n, circuit_duration - t1))
+        .collect();
```

Comment:
You could avoid this step if you're going to convert to a `PyDict` anyway. But if that's the case you should probably consider having a Rust native version of this function that returns a `HashMap`, and then have the Python version perform the conversion to Python types.